prov/usnic/src/usdf_cq.c

```diff
@@ -149,12 +149,122 @@ usdf_cq_readerr_soft(struct fid_cq *fcq, struct fi_cq_err_entry *entry,
 	return 1;
 }
 
+static inline ssize_t
+usdf_cq_copy_cq_entry(void *dst, struct usd_completion *src,
+			enum fi_cq_format format)
+{
+	struct fi_cq_entry *ctx_entry;
+	struct fi_cq_msg_entry *msg_entry;
+	struct fi_cq_data_entry *data_entry;
+
+	switch (format) {
+	case FI_CQ_FORMAT_CONTEXT:
+		ctx_entry = (struct fi_cq_entry *)dst;
+		ctx_entry->op_context = src->uc_context;
+		break;
+	case FI_CQ_FORMAT_MSG:
+		msg_entry = (struct fi_cq_msg_entry *)dst;
+		msg_entry->op_context = src->uc_context;
+		msg_entry->flags = usdf_cqe_to_flags(src);
+		msg_entry->len = src->uc_bytes;
+		break;
+	case FI_CQ_FORMAT_DATA:
+		data_entry = (struct fi_cq_data_entry *)dst;
+		data_entry->op_context = src->uc_context;
+		data_entry->flags = usdf_cqe_to_flags(src);
+		data_entry->len = src->uc_bytes;
+		data_entry->buf = 0; /* XXX */
+		data_entry->data = 0;
+		break;
+	default:
+		USDF_WARN("unexpected CQ format, internal error\n");
+		return -FI_EOPNOTSUPP;
+	}
+
+	return FI_SUCCESS;
+}
+
+static inline ssize_t
+usdf_cq_sread_common(struct fid_cq *fcq, void *buf, size_t count, const void *cond,
+			int timeout, enum fi_cq_format format)
+{
+	struct usdf_cq *cq;
+	uint8_t *entry;
+	uint8_t *last;
+	size_t entry_len;
+	ssize_t ret;
+	size_t time_spent = 0;
+
+	cq = cq_ftou(fcq);
+	if (cq->cq_comp.uc_status != 0)
+		return -FI_EAVAIL;
+
+	switch (format) {
+	case FI_CQ_FORMAT_CONTEXT:
+		entry_len = sizeof(struct fi_cq_entry);
+		break;
+	case FI_CQ_FORMAT_MSG:
+		entry_len = sizeof(struct fi_cq_msg_entry);
+		break;
+	case FI_CQ_FORMAT_DATA:
+		entry_len = sizeof(struct fi_cq_data_entry);
+		break;
+	default:
+		return 0;
+	}
+
+	ret = 0;
+	entry = buf;
+	last = entry + (entry_len * count);
+
+	while (entry < last) {
+		ret = usd_poll_cq(cq->c.hard.cq_cq, &cq->cq_comp);
+		if (ret == -EAGAIN) {
+			if (entry > (uint8_t *)buf)
+				break;
+			if (timeout >= 0 && time_spent >= timeout)
+				break;
+			usleep(1000 * SREAD_SLEEP_TIME_MS);
+			time_spent += SREAD_SLEEP_TIME_MS;
+			continue;
+		}
+		if (cq->cq_comp.uc_status != 0)
+			return -FI_EAVAIL;
```

Comment:
I think this case isn't quite right.  If `entry > buf` and `uc_status != 0` then we will end up returning `-FI_EAVAIL` without indicating how many valid entries are now sitting in `buf`, so the app will essentially "lose" those entries.  I think that if you hit an error you should always return at that point, returning `-FI_EAVAIL` if `entry == buf` or the number of valid entries otherwise.